VirtualKeyboard: Revamp visibility handling

NiLuJe opened the pull request:
Move as much of the state tracking as possible inside VirtualKeyboard itself.
InputDialog unfortunately needs an internal tracking of this state because it needs to know about it *before* the VK is shown, so we have to keep a bit of duplication in there, although we do try much harder to keep everything in sync (at least at function call edges), and to keep the damage contained to, essentially, the toggle button's handler.

(Followup to #10803 & #10850)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/10852)
<!-- Reviewable:end -->

poire-z reviewed (commented):
More green than red, and not sure it's more readable with the many wrappers/forwarders.
But if it makes you happier, nothing against.
Will test, at least one issue noticed, see followup post.

poire-z commented:
A picture with a thousand words should be worth a thousand words (but not really, as its text won't be findable via Search):
![image](https://github.com/koreader/koreader/assets/24273478/8a12fdfc-f967-4284-8a05-6464fc395842)
That is on the emulator (not sure other devices can have both input methods).

poire-z commented:
Some issue with Text editor and just taping twice on the toggle keyboard nav_bar button:

Keyboard (original I guess) shown, but it hides the nav_bar&co buttons (which must still be at bottom), + inputtext height sizing/redrawing issue, showing some old bits of the previous painting:

![image](https://github.com/koreader/koreader/assets/24273478/1c6916d7-27d9-4bcd-aa09-a088f5f3d5d8)

Just typing something (with VK or my PC keyboard) draws the nav bar buttons in that space back. But only once: hit that KB-toggle button again twice, same situation as on the screenshot, but typing does not help this time.

Some other drawing issue when I typed <kbd>Esc</kbd> multiple time to close KOReader :):
![image](https://github.com/koreader/koreader/assets/24273478/b65ff228-aa82-4d7a-bdaa-a462569c16b0)


Somehow, no such issue at all when in Book style tweak (some (the?) other user of the full screen InputDialog):
![image](https://github.com/koreader/koreader/assets/24273478/983bae55-f1df-45e4-a197-298b4d3da1a2)

NiLuJe commented on frontend/ui/widget/inputdialog.lua:
```diff
@@ -475,12 +476,11 @@ function InputDialog:getAddedWidgetAvailableWidth()
 function InputDialog:onTap()
+    -- This is slightly more fine-grained than VK's own visibility lock, hence the duplication...
     if self.deny_keyboard_hiding then
         return
     end
-    if self._input_widget.onCloseKeyboard then
-        self._input_widget:onCloseKeyboard()
-    end
+    self:onCloseKeyboard()
 end
```
I haven't traced that one precisely, but that sounds right (the handler is defined a few lines above that, and it indeed sets the range as "not the keybnoard" ;p).

The `InputText:onTapTextBox` thing is commented further down, though.

NiLuJe commented:
> That is on the emulator (not sure other devices can have both input methods).

Yeah, I was wondering whether the SDL stuff was going to take it in stride, guess not ;p.

NiLuJe commented:
> Some issue with Text editor and just taping twice on the toggle keyboard nav_bar button

Hmm, I tested on a mostly blank file, but, ditto, was hoping it would Just Work there ;p.

NiLuJe commented:
> > 
> 
> Hmm, I tested on a mostly blank file, but, ditto, was hoping it would Just Work there ;p.

Okay, that's sneakier than that, you need to have scrolled for it to screw up.

NiLuJe commented:
Fixed the refresh/scrolling thing, but there was an (existing, AFAICT) weird behavior in that a toggle would restore out of date caret positions (i.e., the last init's, vs., the current), which means you essentially jump around the document when toggling.

I fixed that, but I'm vaguely wondering if that was somehow on purpose? (That feels weird, because there's a lot of engineering going on in that codepath...).

NiLuJe commented:
> No issue in the Add OPDS catalog (MultiInputDialog I guess)

No issues with a simple InputDialog either (e.g., Dic/Wik lookup).

NiLuJe commented:
> No issues with a simple InputDialog either (e.g., Dic/Wik lookup).

So, if I don't forget to apply the patch, yeah, that's reproducible there, which makes this easier to follow ;D.

poire-z commented:
> Looking at RBM

IDK WYM :/

> weird behavior in that a toggle would restore out of date caret positions (i.e., the last init's, vs., the current), which means you'd essentially jump around the document when toggling.
I fixed that, but I'm vaguely wondering if that was somehow on purpose? (That feels weird, because there's a lot of engineering going on in that codepath...).

I would think, given the numerous equivalent of your `self:resyncPos()` in InputText, that we always keep it updated in there - and InputDialog would not need to care or call `self:resyncPos()` and self._input_widget.top_line_num&charpos should be accurate) - so may be I forgot to do it in some place?

> Toggle's SDL text input mode directly inside VK
Ensure it's actually only on when a VK is shown, no matter how that VK
gets shown/hidden.

(When testing this PR, I thought it would actually be nice if the PC keyboard would work as input when the VK keyboard is hidden - no space wasted by the unused VK - but may be it would conflict with keys used for other actions.)
(Or @Frenzie can just toggle the keyboard when his cats pat on the PC keyboard without losing his current edit sessions, so it has its use case!)

poire-z commented:
> weird behavior in that a toggle would restore out of date caret positions (i.e., the last init's, vs., the current), which means you'd essentially jump around the document when toggling.

Can't reproduce this. What's your workflow?

NiLuJe commented:
> Can't reproduce this. What's your workflow?

Text editor (with or without keyboard shown on startup, doesn't matter), scroll via a swipe inside the text (that may be important), toggle keyboard => will jump to the starting position.



> I would think, given the numerous equivalent of your `self:resyncPos()` in InputText, that we always keep it updated in there - and InputDialog would not need to care or call `self:resyncPos()` and self._input_widget.top_line_num&charpos should be accurate) - so may be I forgot to do it in some place?

I would assume so, IIRC, most of the calls inside InputText are for actual (physical) keyboard trigger (and possibly actual on-Kobo physical keyboard, not sure if SDL take these cdepaths).



> When testing this PR, I thought it would actually be nice if the PC keyboard would work as input when the VK keyboard is hidden - no space wasted by the unused VK - but may be it would conflict with keys used for other actions.

The thought crossed my mind, and then I realized I have no idea *why* we even disable SDL input at all. Binding conflicts is probably it? (Fun fact: you can still do *some* stuff when SDL text input is disabled, e.g., enter & return for sure, I assume directional arrows and stuff too, probably?).

NiLuJe commented:
> > Looking at RBM
> 
> IDK WYM :/

ReaderBookMark, but that was a red herring anyway, since I hadn't applied the changes at the time ;D.

NiLuJe commented:
> I would assume so, IIRC, most of the calls inside InputText are for actual (physical) keyboard trigger (and possibly actual on-Kobo physical keyboard, not sure if SDL takes these codepaths).

Meaning most stuff actually goes one or two layer deeper into InputDialog's InputText's ScrollTextWidget's own TextBoxWidget's scroll method. So AFAICT InputDialog's InputText's has no clue about the updated positions, we have to ask it to query its own nested widgets.

Frenzie commented:
> The thought crossed my mind, and then I realized I have no idea why we even disable SDL input at all. Binding conflicts is probably it?

In which context? I thought we'd been working on *adding* it, not disabling it.

NiLuJe commented:
> In which context?

Basically, as it stands, the intent seemed to be to enabled SDL text input as long as a *VirtualKeyboard* is *visible*. This leads to the aforementioned quirks when you have an input field but a VK hidden.

We *could* instead bracket those calls to make it so it's available as long as an *InputDialog* is visible, which would possibly be slightly less confusing, but would introduce another quirk: you'd be able to input text in a potentially unfocused field (likely the last one in focus).

I haven't changed the behavior yet, but added a comment about it.

NiLuJe commented:
> We _could_

Except things get gnarly when you start stacking multiple InputDialog, so, maybe scratch that thought ;p.

Frenzie commented:
> This leads to the aforementioned quirks when you have an input field visible but a VK hidden.

What quirk? VK not visible = field not focused = why would you expect keyboard input to work.

NiLuJe commented:
I mean the fact that navigation still works in that case ;).

poire-z commented on frontend/ui/widget/inputdialog.lua:
```diff
@@ -635,7 +640,7 @@ function InputDialog:onClose()
     self._charpos = self._input_widget.charpos
     if self.view_pos_callback then
-        -- Give back top line num and cursor position
+        -- Let the widget store the current top line num and cursor position
```
"the widget" in all the comments in this commit feels a bit confusing. For me, a "widget" is some UI small element (ie. InputDialog, InfoMessage). I think it would at least need "upper" or "parent": "upper widget".
And widgets don't usually handle/store settings, so "application", "using code", "upper code", something like that would sound better.

NiLuJe replied:
Agreed, and doubly so because the widget is actually, well, *this* widget ;D.

Frenzie commented:
> I meant the fact that navigation w/ the physical keyboard (especially enter & return) still works in that case ;).

Those are implemented very differently (i.e., identical to Kindle with hw kb).

If you meant from a user perspective, sure, that's all missing functionality, but from a code perspective I don't know of any quirks unless they were introduced recently.

NiLuJe commented:
> so...: why do they work ?

Because they're standard keycodes in our mappings, I assume.

i.e., 

> Those are implemented very differently (i.e., identical to Kindle with hw kb).

----

> If it's needed only on save/re-init, may be the one you added in :onClose() is just enough, and we could remove all the others in the little methods ?

I vaguely recall needing them when testing physical keyboards, possibly because the very thing might trigger an immediate scroll, requiring up-to-date coordinates? (It's been a while).

Out of curiosity, *how* were you scrolling in your testing?

poire-z commented:
> Out of curiosity, how were you scrolling in your testing?

Swiping over the textwidget with the mouse on the emulator.
But same behaviour if scrolling with PgUp and PgDown keyboard keys: after hide & show keyboard, we're back to the initial - or last one we went to with cursor keys or text input - position.

NiLuJe commented:
> Can't check right now, need to charge an OTG-capable device ;).

Well, it was charged ;).

Verdict: I don't *quite* remember what they're here for, because at a quick glance I don't break anything by removing them, but, in any case, this does *not* take the InputDialog close/init codepaths (which makes sense, we're just scrolling), soooo, not going to rock the boat too much ;p.

poire-z commented:
Quickly re-tested, on the emu and my Kobo: no issue noticed.
(Quickly checked on the emulator that TerminalEmulator, using another inputtext_class than InputText, seems to work fine, but not a real user of this plugin.)

I thought that, now that (I think?) you just hide/show a same keyboard instead of maybe recreating one each time you hide/show, that the keyboard layout we were in when hiding would be the same when shown again - but apparently not (which is fine by me :).

> I don't break anything by removing them, but, in any case, this does not take the InputDialog close/init codepaths (which makes sense, we're just scrolling), soooo, not going to rock the boat too much ;p.

All ok with not rocking the boat :) (but this is just so not you :))
(Moving cursor up/down/left/right is also just scrolling.
And if charpos/top_line_num were to be really needed in other situations than re-init, we would meet the same bug you noticed when one scroll with swipe and charpos/top_line_num being innacurate.
It's also possible upper widget may peek charpos/top_line_num (didn't find any occurence, but I think I would not have hesitated peeking them from an upper widget if I needed them.
)

NiLuJe commented:
> I thought that, now that (I think?) you just hide/show a same keyboard instead of maybe recreating one each time you hide/show, that the keyboard layout we were in when hiding would be the same when shown again - but apparently not (which is fine by me :).

Yeah, that'd be a whole 'nother kettle of fish ;p.

(And a bit of a fool's errand when the rest of the input widgets are always destroyed and recreated from scratch at runtime anyway).


> All ok with not rocking the boat :) (but this is just so not you :))

;D. FWIW, I checked the blame, and I must have been thinking of some other small positioning related thingy when I was reminiscing about the OTG keyboard testing, because those are much, much older than that :s.

NiLuJe reviewed (commented):
> Reviewed 2 of 6 files at r1, 1 of 2 files at r4, 2 of 4 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 1 of 2 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/10852)* status: all files reviewed, 2 unresolved discussions (waiting on @poire-z)



<!-- Sent from Reviewable.io -->
